editor/src/messages/tool/common_functionality/auto_panning.rs

```diff
@@ -86,3 +86,302 @@ impl AutoPanning {
+		let initial_transform = editor.active_document().metadata().transform_to_viewport(layer);
+
+		editor.select_tool(ToolType::Select).await;
+		editor.left_mousedown(150., 150., ModifierKeys::empty()).await; // Click near edge for resize handle
+		editor
+			.mouseup(
+				EditorMouseState {
+					editor_position: DVec2::new(150., 150.),
+					mouse_keys: MouseKeys::empty(),
+					scroll_delta: ScrollDelta::default(),
+				},
+				ModifierKeys::empty(),
+			)
+			.await;
+
+		editor.handle_message(TransformLayerMessage::BeginScale).await;
+
+		// Moving cursor to trigger auto-panning
+		editor.move_mouse(2000., 2000., ModifierKeys::empty(), MouseKeys::LEFT).await;
+
+		let pointer_keys = SelectToolPointerKeys {
+			axis_align: Key::Shift,
+			snap_angle: Key::Control,
+			center: Key::Alt,
+			duplicate: Key::Alt,
+		};
+
+		// Simulatiing auto-panning for several frames
+		for _ in 0..5 {
+			editor.handle_message(SelectToolMessage::PointerOutsideViewport(pointer_keys.clone())).await;
+		}
+
+		editor.handle_message(TransformLayerMessage::ApplyTransformOperation { final_transform: true }).await;
+
+		// Verifying the transform has changed (scale component should be different)
+		let final_transform = editor.active_document().metadata().transform_to_viewport(layer);
+
+		// Comparing transform matrices to detect scale changes
+		let initial_scale = initial_transform.matrix2.determinant().sqrt();
+		let final_scale = final_transform.matrix2.determinant().sqrt();
+		let scale_ratio = final_scale / initial_scale;
+
+		assert!(
+			scale_ratio > 1.1 || scale_ratio < 0.9,
+			"Rectangle should have been resized due to auto-panning (scale ratio: {})",
+			scale_ratio
+		);
+	}
+
+	#[tokio::test]
+	async fn test_artboard_tool_drawing_auto_panning() {
+		let mut editor = EditorTestUtils::create();
+		editor.new_document().await;
+
+		editor.select_tool(ToolType::Artboard).await;
+		editor.left_mousedown(100., 100., ModifierKeys::empty()).await;
+
+		// Moving cursor outside viewport to trigger auto-panning
+		editor.move_mouse(2000., 100., ModifierKeys::empty(), MouseKeys::LEFT).await;
+
+		// Sending pointer outside event multiple times to simulate auto-panning
+		for _ in 0..5 {
+			editor
+				.handle_message(ArtboardToolMessage::PointerOutsideViewport {
+					constrain_axis_or_aspect: Key::Shift,
+					center: Key::Alt,
+				})
+				.await;
+		}
+
+		editor
+			.mouseup(
+				EditorMouseState {
+					editor_position: DVec2::new(2000., 100.),
+					mouse_keys: MouseKeys::empty(),
+					scroll_delta: ScrollDelta::default(),
+				},
+				ModifierKeys::empty(),
+			)
+			.await;
+
+		// Verifying that an artboard was created with significant width due to auto-panning
+		let artboards = get_artboards(&mut editor).await;
+		assert!(!artboards.is_empty(), "Artboard should have been created");
+		assert!(
+			artboards[0].dimensions.x > 200,
+			"Artboard should have significant width due to auto-panning: {}",
```

Comment:
What led you to decide that this was «significant». Please justify in a comment.